lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,818 @@
+//! #         &self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+//! #     ) -> Result<PaymentId, PaymentSendFailure> { unimplemented!() }
+//! #     fn retry_payment(
+//! #         &self, route: &Route, payment_id: PaymentId
+//! #     ) -> Result<(), PaymentSendFailure> { unimplemented!() }
+//! # }
+//! #
+//! # struct FakeRouter {};
+//! # impl Router for FakeRouter {
+//! #     fn find_route(
+//! #         &self, payer: &PublicKey, payee: &PublicKey, payee_features: Option<InvoiceFeatures>,
+//! #         first_hops: Option<&[&ChannelDetails]>, last_hops: &[&RouteHint],
+//! #         final_value_msat: u64, final_cltv: u32
+//! #     ) -> Result<Route, LightningError> { unimplemented!() }
+//! # }
+//! #
+//! # struct FakeLogger {};
+//! # impl Logger for FakeLogger {
+//! #     fn log(&self, record: &Record) { unimplemented!() }
+//! # }
+//! #
+//! # fn main() {
+//! let event_handler = |event: &Event| {
+//!     match event {
+//!         Event::PaymentPathFailed { .. } => println!("payment failed after retries"),
+//!         Event::PaymentSent { .. } => println!("payment successful"),
+//!         _ => {},
+//!     }
+//! };
+//! # let payer = FakePayer {};
+//! # let router = FakeRouter {};
+//! # let logger = FakeLogger {};
+//! let invoice_payer = InvoicePayer::new(&payer, router, &logger, event_handler)
+//!     .with_retry_attempts(2);
+//!
+//! let invoice = "...";
+//! let invoice = invoice.parse::<Invoice>().unwrap();
+//! invoice_payer.pay_invoice(&invoice).unwrap();
+//!
+//! # let event_provider = FakeEventProvider {};
+//! loop {
+//!     event_provider.process_pending_events(&invoice_payer);
+//! }
+//! # }
+//! ```
+//!
+//! # Note
+//!
+//! The [`Route`] is computed before each payment attempt. Any updates affecting path finding such
+//! as updates to the network graph or changes to channel scores should be applied prior to
+//! retries, typically by way of composing [`EventHandler`]s accordingly.
+
+use crate::Invoice;
+
+use bitcoin_hashes::Hash;
+use bitcoin_hashes::sha256::Hash as Sha256;
+
+use lightning::ln::{PaymentHash, PaymentSecret};
+use lightning::ln::channelmanager::{ChannelDetails, PaymentId, PaymentSendFailure};
+use lightning::ln::features::InvoiceFeatures;
+use lightning::ln::msgs::LightningError;
+use lightning::routing::router::{Route, RouteHint};
+use lightning::util::events::{Event, EventHandler};
+use lightning::util::logger::Logger;
+
+use secp256k1::key::PublicKey;
+
+use std::collections::hash_map::{self, HashMap};
+use std::ops::Deref;
+use std::sync::Mutex;
+
+/// A utility for paying [`Invoice]`s.
+pub struct InvoicePayer<P: Deref, R, L: Deref, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	payer: P,
+	router: R,
+	logger: L,
+	event_handler: E,
+	invoice_cache: Mutex<HashMap<PaymentHash, (Invoice, usize)>>,
+	retry_attempts: usize,
+}
+
+/// A trait defining behavior of an [`Invoice`] payer.
+pub trait Payer {
```

Comment:
We need some way to serialize a `Payer` to disk or otherwise store it in a user-provided K-V store. I think we should require it be persistent in the way that `ChannelMonitor`s are persistent, ie we have some function which we call during operation which requires the user persist an updated `Payer` state to disk before the `Payer` take action.

I think this will work out well -
 * in `pay_invoice` either we
    * persist the `Payer` with a new payment but the `ChannelManager`/`ChannelMonitor` does not update, then we won't have returned to the user from `pay_invoice`, and they should know to retry because they never finished the method call. Somewhat annoyingly, on restart, we'll refuse to make the payment even though its not actually in-flight. I'm not entirely sure how to handle this without interrogating `ChannelMonitor`, but at least the payment will just be "pending" forever and not "success" or "failed".
    * persist the `Payer` with an new payment and the `ChannelMonitor`, but no `ChannelManager` update. This may imply we've returned success from `pay_invoice` before shutdown/crash. In this case on restart the `ChannelManager` will close the channel due to being out of sync with the monitor, and the `ChannelMonitor` will tell us the final state of the HTLC which will generate a payment even for the `Payer` to handle, even though the `ChannelManager` has no idea what is going on. This will require an update to `ChannelManager`, see #1102.
    * persist all three, in which case we'll handle stuff well (duh).

   We will need to update the documentation accordingly here, noting that once you call `send_payment`, literally even if it doesn't return back to user code, you must consider the payment in-flight!
 * in payment retrying, things are a bit simpler, because we know the `ChannelManager` wont be re-persisted until we return so either we process the event or the manager won't be updated. We could play the event twice, however, so maybe we *do* need to store the pending paths in the `Payer`? Ugh.